File: leetcode_ly/c++/BackTracking/PathSumTwo.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> result;
        vector<int> ans;
        getResult(root,result,ans,sum);
        return result;
    }
    
    void getResult(TreeNode* root,vector<vector<int>>& result,vector<int>& ans,int sum){
        if(!root) return;
        ans.push_back(root->val);
        if(!root->left && !root->right && sum == root->val){
            result.push_back(ans);
            //return;
        }

        getResult(root->left,result,ans,sum - root->val);
        getResult(root->right,result,ans,sum - root->val);
        ans.pop_back();
        return;
    }
};

File: leetcode_ly/java/Tree/PathSumTwo.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> list = new ArrayList<List<Integer>>();
        if(root == null) return list;
        List<Integer> inList = new ArrayList<Integer>();
        getPath(root,sum,list,inList);
        return list;
    }
    
    public void getPath(TreeNode root,int sum,List<List<Integer>> list,List<Integer> inList){
        if(root == null) return;
        inList.add(root.val);
        if(root.left == null && root.right == null && root.val == sum){
            list.add(new ArrayList<Integer>(inList));
        }
        
        getPath(root.left,sum - root.val,list,inList);
        getPath(root.right,sum - root.val,list,inList);
        inList.remove(inList.size() - 1);
        return;
    }
}